### PR TITLE
Workflow delete

### PR DESCRIPTION
**Pull request by emilytoppm**
- Adds the ability to deactivate workflows and tasks, which cancels any in progress states

### CODE REVIEW

File: wagtail/admin/templates/wagtailadmin/workflows/confirm_disable.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{% extends "wagtailadmin/base.html" %}
+{% load i18n %}
+{% block titletag %}{{ view.page_title }}{% endblock %}
+
+{% block content %}
+    {% include "wagtailadmin/shared/header.html" with title=view.page_title subtitle=view.get_page_subtitle icon=view.header_icon %}
+
+    <div class="nice-padding">
+        {% if workflow_states_in_progress %}
+            <p>This workflow is currently in progress on {{ workflow_states_in_progress }} pages. Disabling this workflow will cancel moderation on these page.</p>
```
Comment (gasman) [inline wagtail/admin/templates/wagtailadmin/workflows/confirm_disable.html]: "these page" -> "these pages" - we should really handle the singular vs plural through the translations framework though.

File: wagtail/admin/edit_handlers.py [hunk 2]
```diff
@@ -789,19 +789,40 @@ def __init__(self, **kwargs):
...
 Task.base_form_class = WagtailAdminModelForm
 
 
+class ExcludeFieldsOnEditMixin:
+    """A mixin for edit handlers, which disables fields listed in a model's 'exclude_on_edit' attribute when binding
+    to an existing instance - editing rather than creating"""
+
+    def bind_to(self, *args, **kwargs):
+        new = super(ExcludeFieldsOnEditMixin, self).bind_to(*args, **kwargs)
+        # when binding to an existing instance - ie editing - set those fields in the form to disabled
```
Comment (gasman) [inline wagtail/admin/edit_handlers.py]: I don't think we can assume that the presence of an instance means that we're editing - for example, the create page view also binds to a newly-created instance: https://github.com/wagtail/wagtail/blob/2e23ef4ee45ee553d0f1647a7a2ecb09983a19a3/wagtail/admin/views/pages.py#L216

(Perhaps checking `new.instance.pk is not None` would be sufficient to fix this?)

File: wagtail/admin/templates/wagtailadmin/workflows/edit.html [hunk 3]
```diff
@@ -34,14 +34,26 @@
...
                         </button>
-                        {% if can_delete %}
-                        <a href="{{ view.get_delete_url }}" class="button button-secondary no">{{ view.delete_item_label }}</a>
-                    {% endif %}
                     </p>
                 </li>
             </ul>
         </div>
     </form>
+    {% if can_enable or can_delete %}
```
Comment (gasman) [inline wagtail/admin/templates/wagtailadmin/workflows/edit.html]: `can_delete` -> `can_disable`

File: wagtail/admin/templates/wagtailadmin/workflows/edit_task.html [hunk 4]
```diff
@@ -43,4 +43,19 @@
             </footer>
         {% endblock %}
     </form>
+    {% if can_enable or can_delete %}
```
Comment (gasman) [inline wagtail/admin/templates/wagtailadmin/workflows/edit_task.html]: `can_delete` -> `can_disable`

File: wagtail/admin/views/workflows.py [hunk 5]
```diff
@@ -192,6 +247,17 @@ class TaskIndex(IndexView):
     add_item_label = _("Add a task")
     header_icon = 'placeholder'
 
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        if not strtobool(self.request.GET.get('show_disabled', 'False')):
```
Comment (gasman) [inline wagtail/admin/views/workflows.py]: I'm not really keen on capitalised parameters in URLs, and importing `strtobool` from distutils is a bit leftfield :-) The standard elsewhere for boolean URL parameters is for the presence of a parameter (with any value) to mean true, so this becomes `if not self.request.GET.get('show_disabled')`.

File: wagtail/admin/templates/wagtailadmin/workflows/index.html [hunk 6]
```diff
@@ -58,9 +58,17 @@ <h1 class="icon icon-{{ view.header_icon }}">
...
             {% url view.add_url_name as add_url %}
-            <p>{% blocktrans %}No workflows have been created. Why not <a href="{{ add_url }}">add one</a>?{% endblocktrans %}</p>
+            {% if showing_disabled %}
+                <p>{% blocktrans %}No workflows have been created. Why not <a href="{{ add_url }}">add one</a>?{% endblocktrans %}</p>
+            {% else %}
+                <p>{% blocktrans %}There are no enabled workflows. Why not <a href="{{ add_url }}">add one</a>?{% endblocktrans %}</p>
+            {% endif %}
+        {% endif %}
+        {% if showing_disabled %}
+            <a href="?show_disabled=False" class="button secondary">{% trans "Hide Disabled Workflows" %}</a>
```
Comment (gasman) [inline wagtail/admin/templates/wagtailadmin/workflows/index.html]: Standard convention for button text is to only capitalise the first word - "Hide disabled workflows". (It gets capitalised in CSS anyway, but at least this way we keep it consistent for translators...)

**Review by gasman — CHANGES_REQUESTED** [review]

File: wagtail/admin/templates/wagtailadmin/workflows/index.html (hunk 6 above)
Comment (emilytoppm) [reply wagtail/admin/templates/wagtailadmin/workflows/index.html]: Thanks, I've fixed this now.

File: wagtail/admin/views/workflows.py (hunk 5 above)
Comment (emilytoppm) [reply wagtail/admin/views/workflows.py]: Oops, sorry about that, didn't know the convention for that - I've amended now to just checking for the key in `request.GET`

File: wagtail/admin/templates/wagtailadmin/workflows/edit_task.html (hunk 4 above)
Comment (emilytoppm) [reply wagtail/admin/templates/wagtailadmin/workflows/edit_task.html]: Thanks, fixed!

File: wagtail/admin/templates/wagtailadmin/workflows/edit.html (hunk 3 above)
Comment (emilytoppm) [reply wagtail/admin/templates/wagtailadmin/workflows/edit.html]: Thanks, fixed!

File: wagtail/admin/edit_handlers.py (hunk 2 above)
Comment (emilytoppm) [reply wagtail/admin/edit_handlers.py]: Oh yes, good point - it worked for Tasks, but you're right that it wouldn't work elsewhere, thanks. I've added an additional check for `new.instance.pk is not None`

File: wagtail/admin/templates/wagtailadmin/workflows/confirm_disable.html (hunk 1 above)
Comment (emilytoppm) [reply wagtail/admin/templates/wagtailadmin/workflows/confirm_disable.html]: Ah thanks - I've changed this over to use `ngettext` now, let me know if that's the translation approach you were thinking of?

Comment (gasman) [reply wagtail/admin/templates/wagtailadmin/workflows/confirm_disable.html]: The `{% blocktrans %}` tag supports a `{% plural %}` sub-tag which would avoid the need to move the text out of the template - either is good though :-)

**Review by gasman — APPROVED** [review]

**Pull request merged**